app/src/main/java/org/kiwix/kiwixmobile/di/modules/KiwixModule.kt

```diff
@@ -52,4 +53,9 @@ object KiwixModule {
       sharedPreferenceUtil,
       listOf(MountFileSystemChecker(mountPointProducer), FileWritingFileSystemChecker())
     )
+
+  @Provides
+  @KiwixScope
+  fun providesWiFiP2pManager(context: Context): WifiP2pManager =
+    context.getSystemService(Context.WIFI_P2P_SERVICE) as WifiP2pManager
```

Comment:
well this won't pass api 27

Reply:
Yeah, but when I initialize via assuring nullability, it doesn't let me do so. This format also doesn't work
``` Kotlin 
private val manager: WifiP2pManager? by lazy(LazyThreadSafetyMode.NONE) {
    activity.getSystemService(Context.WIFI_P2P_SERVICE) as WifiP2pManager?

Reply:
We can't assure this is non null, that is the point of the error.

````
  @Provides
  @KiwixScope
  fun providesWiFiP2pManager(context: Context): WifiP2pManager? =
    context.getSystemService(Context.WIFI_P2P_SERVICE) as? WifiP2pManager
```

```
class WifiDirectManager @Inject constructor(
  private val activity: Activity,
  private val sharedPreferenceUtil: SharedPreferenceUtil,
  private val alertDialogShower: AlertDialogShower,
  private val manager: WifiP2pManager?
) 
```

the  return type here 
` fun providesWiFiP2pManager(context: Context): WifiP2pManager?`

has to match the type here

`  private val manager: WifiP2pManager?`

because dagger goes "I've been asked to inject a `WifiP2pManager?` do I have a recipe for this? oh yup look at this return type in KiwixModule"

Reply:
Oho, so only mistake I'm committing was the asserting the `manager` non-nullable in the `directmanager` file, gotcha. But why dagger is so specific? It could've simply injected it, although the only difference was of nullability.

Reply:
Dagger works on types, `WifiDirectManager` != `WifiDirectManager?` and the error still persists because `context.getSystemService(Context.WIFI_P2P_SERVICE)` returns null on 27 and you try to cast it to a non nullable type

Reply:
It works cool now, what change led you to suspect this?

Reply:
To what are you referring? `context.getSystemService(Context.WIFI_P2P_SERVICE)` returning null? Because when you cast it to the nullable type the tests on 27 passed, it being lazy had nothing to do with it working and only the fact of the nullable cast ensured that.

Reply:
Yeah, I totally agree with the fact that lazy did nothing but the cast did the work. I was referring to the test that was failing on API 27

Reply:
What led me to suspect the test was failing on api 27? Is that the question? Feel free to message me if that'll more quickly answer your question

Reply:
Fine